cartographer/mapping/pose_graph/cost_helpers_impl.h

```diff
@@ -92,24 +92,27 @@ std::array<T, 6> ScaleError(std::array<T, 6> error, T translation_weight,
   return scaled_error;
 }
 
+//  Eigen implementation of slerp is not compatible with Ceres on all supported
+//  platforms. Our own implementation is used instead.
 template <typename T>
 std::array<T, 4> SlerpQuaternions(const T* const start, const T* const end,
                                   T factor) {
   // Angle 'theta' is the half-angle "between" quaternions. It can be computed
   // as the arccosine of their dot product.
   const T cos_theta = start[0] * end[0] + start[1] * end[1] +
                       start[2] * end[2] + start[3] * end[3];
-  // If numerical error brings 'cos_theta' outside of [-1., 1.] interval, then
-  // the quaternions are likely to be collinear.
-  if (cos_theta >= T(1.0) || cos_theta <= T(-1.0)) {
-    return {{end[0], end[1], end[2], end[3]}};
+  const T abs_cos_theta = abs(cos_theta);
+  // If numerical error brings 'cos_theta' outside [-1 + epsilon, 1 - epsilon]
+  // interval, then the quaternions are likely to be collinear.
+  T prev_scale = T(1.) - factor;
+  T next_scale = factor;
+  if (abs_cos_theta < T(1. - 1e-5)) {
```

Comment:
Looks good.

Reply:
Thanks for noticing this division by 0.

Reply:
Interesting that Eigen does not handle this case specially https://bitbucket.org/eigen/eigen/src/2355b229ea4c2876f490e726526cdcd8a63c7f54/Eigen/src/Geometry/Quaternion.h?at=default&fileviewer=file-view-default#Quaternion.h-748